---
title: "Fst_01172022"
author: "Sophia Buysse"
date: "1/17/2022"
output: html_document
---
Need document with less strict filtering parameters... why not just use the allsites one? try that first to see if it works honestly.
script is called Fst_sub_SNPs_2022.R for the January version
Fst_sub_2022 for the May 2022 version

load R/4.2.2 so the version of R that I am loading matches the version of R that I am running on both my laptop and my lab PC

module purge
module load  GCC/11.2.0  OpenMPI/4.1.1 R/4.2.2
R

scratch that! can't get diveRisty to install on that version on the HPCC???? why I don't know, but as a result I'm using version R/4.1.2 on the HPCC because I know that one works at least.

The install error seems to be something with GNU and there is something required in my linux background that includes libtoolize but I haven't found a way to get that without making R/4.2.2 unable to be loaded


Population Pairwise Fst was cut from the manuscript because it reports similar information to the genetic PCA and we had concerns that the patterns were artifacts of sequencing rather than reliable results of differences between populations.


# Plink Filtered Files

## With centromeres
```{r pairwise_fst_HPCC, eval = FALSE}
library(dplyr)
library(tidyr)
library(ape)

library(diveRsity)
#library(pegas)

library(genio)

##### genio to read in data #####

SNPs_Cent <- read_plink("/mnt/scratch/buysseso/GVCF/allsites_filtered_plinkTest", verbose = TRUE)

tmp <- geno_to_char(SNPs_Cent$X, SNPs_Cent$bim)

# may want geno_to_char for reformatting data into what I need.

########## Keep Line ID ##########

##### format data #####
colnames(tmp) <- gsub(pattern = "-", replacement = "", colnames(tmp))
colnames(tmp) <- toupper(colnames(tmp))
colnames(tmp)
# make column names all uppercase and remove dashes so all in the same format. Still identifiable by unique plant here, not by population. Might need to change this b/c Fst is between populations.
class(tmp)

# add column of the rownames to function as SNP_ID because the location didn't come through with the vcf
# used cbind!
tmp2 <- as.data.frame(cbind(rownames(tmp),tmp))
colnames(tmp2)[1] <- "SNP_ID"
head(tmp2)
class(tmp2)
#data.frame

# same for rownames
tmp2 <- rbind(colnames(tmp2), tmp2)
head(tmp2)

# lapply! make characters into readable format
#dat2 <- lapply(dat2, chartr, old = "|", new = "/")
tmp2 <- lapply(tmp2, gsub, pattern = "/", replacement = "")

# need to do both of these because some are / and some |
# can't use this gsub for the period because it interprets the . as an anything character
#dat2 <- lapply(dat2, chartr, old = ".", new = "-")
# there was definitely an easier way to do this. oh well!
tmp2 <- as.data.frame(tmp2)
head(tmp2)

save(tmp2, file = "~/R/GenotypeMatrix_plinkFiltered_Cent_Dec2022.ROBJ")
# on 8/3/2023 this output file was copied to /mnt/research/josephslab/Sophie/Qpc

#### subset 10,000 randomly chosen SNPs for fst matrix ####

# this does include the centromeres.
rows_keep <- runif(n = 10000, min = 2, max = length(tmp2$SNP_ID))
rows_keep <- c(1,rows_keep)
# need to keep row one for the ID, set rows_keep to start at two so I never have row 1 twice
# want to keep all columns because each column is an individual, only filter rows
dat_sub <- tmp2[rows_keep, ]
str(dat_sub)
class(dat_sub)
length(dat_sub$SNP_ID)
#10,001

# writes to working directory and I can't change it
# I also can't change the writing name in the output file here. which is not ideal so it is manually changed.
setwd("~/R")
snp2gen(infile = dat_sub, prefix_length = 3)

#clear list to clear up space

rm(list = ls())

### let's try to get some stats out
## I can just start at this point when I come back because it reads in a file that was saved
# BUT note that this will be the last file that was saved so rerunning all the code if coming back on a different day is a good idea

test_output <- fastDivPart(infile = "~/R/snp2gen_converted.gen", outfile = "fastDivPart_plinkFilter_sub_SNPs_Dec16_2022", gp = 2, pairwise = TRUE, fst = TRUE)
# started at 2:17pm on 12/16/2022 for a subset of 10,000 snps
# ended by  2:29pm on 12/16/2022
# this output is an excel file that cannot be looked at with mobaxterm.

save(test_output, file = "~/R/fastDivPart_plinkFilter_sub_SNPS_Dec2022.ROBJ")

########## Only POP ID ##########

##### load data #####
rm(list = ls())
load("~/R/GenotypeMatrix_plinkFiltered_Cent_Dec2022.ROBJ")
# on 8/3/2023 this output file was copied to /mnt/research/josephslab/Sophie/Qpc
##### format data #####
colnames(tmp2)[2:62] <- substr(toupper(colnames(tmp2)[2:62]), 1, 3)
colnames(tmp2)

tmp2[1, ] <- colnames(tmp2)
head(tmp2)

save(tmp2, file = "~/R/GenotypeMatrix_plinkFiltered_Cent_Pop_Dec2022.ROBJ")

#### subset 10,000 randomly chosen SNPs for fst matrix ####

# this does include the centromeres.
rows_keep <- runif(n = 10000, min = 2, max = length(tmp2$SNP_ID))
rows_keep <- c(1,rows_keep)
# need to keep row one for the ID, set rows_keep to start at two so I never have row 1 twice
# want to keep all columns because each column is an individual, only filter rows
dat_sub <- tmp2[rows_keep, ]
str(dat_sub)
class(dat_sub)
length(dat_sub$SNP_ID)
#10,001

# writes to working directory and I can't change it
# I also can't change the writing name in the output file here. which is not ideal so it is manually changed.
setwd("~/R")
snp2gen(infile = dat_sub, prefix_length = 3)

#clear list to clear up space

rm(list = ls())

### let's try to get some stats out
## I can just start at this point when I come back because it reads in a file that was saved
# BUT note that this will be the last file that was saved so rerunning all the code if coming back on a different day is a good idea

test_output <- fastDivPart(infile = "~/R/snp2gen_converted.gen", outfile = "fastDivPart_plinkFilter_sub_SNPs_Pop_Dec16_2022", gp = 2, pairwise = TRUE, fst = TRUE)
# started at 10:12am on 1/6/2023 for a subset of 10,000 snps
# ended at  10:20

save(test_output, file = "~/R/fastDivPart_plinkFilter_sub_SNPS_Pop_Dec2022.ROBJ")


```

## Without Centromeres

```{r pairwise_fst_HPCC2, eval = FALSE}
library(dplyr)
library(tidyr)
library(ape)
library(diveRsity)
library(genio)

##### genio to read in data #####

SNPs_NoCent <- read_plink("/mnt/scratch/buysseso/GVCF/allChrs_SNPs_NoCent", verbose = TRUE)

tmp <- geno_to_char(SNPs_NoCent$X, SNPs_NoCent$bim)

# may want geno_to_char for reformatting data into what I need.

########## Individual ID ###########

##### format data #####
colnames(tmp) <- gsub(pattern = "-", replacement = "", colnames(tmp))
colnames(tmp) <- toupper(colnames(tmp))
colnames(tmp)
# make column names all uppercase and remove dashes so all in the same format. Still identifiable by unique plant here, not by population. Might need to change this b/c Fst is between populations.
class(tmp)


# used cbind to make SNp location a SNP_ID column!
tmp2 <- as.data.frame(cbind(rownames(tmp),tmp))
colnames(tmp2)[1] <- "SNP_ID"
head(tmp2)
class(tmp2)
#data.frame

# same for rownames
tmp2 <- rbind(colnames(tmp2), tmp2)
head(tmp2)

# lapply! make characters into readable format
tmp2 <- lapply(tmp2, gsub, pattern = "/", replacement = "")

tmp2 <- as.data.frame(tmp2)
head(tmp2)

save(tmp2, file = "~/R/GenotypeMatrix_plinkFiltered_NoCent_Dec2022.ROBJ")
# on 8/3/2023 this output file was copied to /mnt/research/josephslab/Sophie/Qpc

#### subset 10,000 randomly chosen SNPs for fst matrix ####

# this does not include the centromeres.
rows_keep <- runif(n = 10000, min = 2, max = length(tmp2$SNP_ID))
rows_keep <- c(1,rows_keep)
# need to keep row one for the ID, set rows_keep to start at two so I never have row 1 twice
# want to keep all columns because each column is an individual, only filter rows
dat_sub <- tmp2[rows_keep, ]
str(dat_sub)
class(dat_sub)
length(dat_sub$SNP_ID)
#10,001

# writes to working directory and I can't change it
# I also can't change the writing name in the output file here. which is not ideal so it is manually changed.
setwd("~/R")
snp2gen(infile = dat_sub, prefix_length = 3)

#clear list to clear up space

rm(list = ls())

### let's try to get some stats out
## I can just start at this point when I come back because it reads in a file that was saved
# BUT note that this will be the last file that was saved so rerunning all the code if coming back on a different day is a good idea

test_output <- fastDivPart(infile = "~/R/snp2gen_converted.gen", outfile = "fastDivPart_plinkFilter_sub_NoCent_SNPs_Jan5_2023", gp = 2, pairwise = TRUE, fst = TRUE)
# started at 9:45am on 1/6/2023 for a subset of 10,000 snps
# ended by  9:56 
# this output is an excel file that cannot be looked at with mobaxterm or a text file, depending on if all the dependencies loaded.

save(test_output, file = "~/R/fastDivPart_plinkFilter_sub_NoCent_SNPS_Dec2022.ROBJ")
#keeping the Dec date just so all the files show up together.

########## Only POP ID ##########

##### load data #####
rm(list = ls())
load("~/R/GenotypeMatrix_plinkFiltered_NoCent_Dec2022.ROBJ")

##### format data #####
colnames(tmp2)[2:62] <- substr(toupper(colnames(tmp2)[2:62]), 1, 3)
colnames(tmp2)

tmp2[1, ] <- colnames(tmp2)
head(tmp2)

save(tmp2, file = "~/R/GenotypeMatrix_plinkFiltered_NoCent_Pop_Dec2022.ROBJ")

#### subset 10,000 randomly chosen SNPs for fst matrix ####

# this does include the centromeres.
rows_keep <- runif(n = 10000, min = 2, max = length(tmp2$SNP_ID))
rows_keep <- c(1,rows_keep)
# need to keep row one for the ID, set rows_keep to start at two so I never have row 1 twice
# want to keep all columns because each column is an individual, only filter rows
dat_sub <- tmp2[rows_keep, ]
str(dat_sub)
class(dat_sub)
length(dat_sub$SNP_ID)
#10,001

# writes to working directory and I can't change it
# I also can't change the writing name in the output file here. which is not ideal so it is manually changed.
setwd("~/R")
snp2gen(infile = dat_sub, prefix_length = 3)

#clear list to clear up space



### let's try to get some stats out
## I can just start at this point when I come back because it reads in a file that was saved
# BUT note that this will be the last file that was saved so rerunning all the code if coming back on a different day is a good idea

test_output <- fastDivPart(infile = "~/R/snp2gen_converted.gen", outfile = "fastDivPart_plinkFilter_sub_NoCent_SNPs_Pop_Jan6_2023", gp = 2, pairwise = TRUE, fst = TRUE)
# started at 10:23am on 1/6/2023 for a subset of 10,000 snps
# ended by  10:31

save(test_output, file = "~/R/fastDivPart_plinkFilter_sub_NoCent_SNPS_Pop_Dec2022.ROBJ")

```

# Visualizing Fst pairwise matrix: diveRsity

This was done on my local machine. Did some manual renaming of files and changing text files to .csv files cause I"m being lazy.

## With centromeres

```{r fst_heatmap, eval = TRUE}
##### visualizing fst matrices
library(RColorBrewer)
library(pheatmap)

##### functions
## to fill in top part of matrix
makeSym <- function(m) {
  m[upper.tri(m)] <- t(m)[upper.tri(m)]
  return(m)
}

##### start with when Fst was calculated with individual identifiers #####

## read in the matrix
sub_10k <- read.csv("C:/Users/Sophie/Michigan State University/Conner, Jeffrey - SophieAnalyses/PopGen/fastDivPart_plinkFilter_sub_SNPs_Dec16_2022/pairwiseFst_WC.csv", header = TRUE)

## reorganize
rownames(sub_10k) <- sub_10k[,1]
sub_10k <- sub_10k[1:16,2:17]
sub_10k <- lapply(sub_10k, as.numeric)
sub_10k <- as.matrix(as.data.frame(sub_10k))
rownames(sub_10k) <- colnames(sub_10k)
# simplify for plotting
rownames(sub_10k) <- substr(rownames(sub_10k), start = 1, stop = 3)
colnames(sub_10k) <- substr(colnames(sub_10k), start = 1, stop = 3)

# get a mean and sd
stats <- as.vector(sub_10k)
sum_fst <- c(mean(stats, na.rm = TRUE), sd(stats, na.rm = TRUE), min(stats, na.rm = TRUE), max(stats, na.rm = TRUE))
# mean= 0.473, sd = 0.169
# min = 0.1057
# max = 0.865

sub_10k <- makeSym(sub_10k)
str(sub_10k)

diag(sub_10k) <- 0

str(sub_10k)
fst_mat <- sub_10k
substr(rownames(fst_mat),1,3)[substr(rownames(fst_mat),1,3) == 'ARB'] <- "ARU"
substr(colnames(fst_mat),1,3)[substr(colnames(fst_mat),1,3) == 'ARB'] <- "ARU"
substr(rownames(fst_mat),1,3)[substr(rownames(fst_mat),1,3) == 'SPE'] <- "SAL"
substr(colnames(fst_mat),1,3)[substr(colnames(fst_mat),1,3) == 'SPE'] <- "SAL"
save(fst_mat, file = "C:/Users/Sophie/Michigan State University/Conner, Jeffrey - SophieAnalyses/Fst_matrix_Dec2022_10k.ROBJ")

pdf("C:/Users/Sophie/Michigan State University/Conner, Jeffrey - SophieAnalyses/Figures/Fst_heatmap_Dec2022_10k.pdf", height = 9, width = 11)
heatmap(x = fst_mat, symm = TRUE,
        xlab = "Population",
        ylab = "Population",
        main = "Fst Heatmap", 
        keep.dendro = FALSE,
        #Rowv = NA,
        col= colorRampPalette(brewer.pal(8, "Oranges"))(255))
legend(x="bottomright", legend=c("0", "0.5", "1"),
     fill=colorRampPalette(brewer.pal(8, "Oranges"))(3))
dev.off()

## add in elevation information
elev.nam <- c('RAB', 'PIN', 'SAL', 'HOR', 'ARU', 'BAR', 'COC', 'POB', 'BOS', 'MUR', 'VDM', 'ALE', 'BIS', 'PAL', 'VIE', 'PAN')
elev <- c(1229, 416, 441, 1444, 715, 519, 413, 836, 1585, 1706, 177, 665, 61, 303, 991, 1605)
elev_mat <- as.data.frame(sub_10k)
colnames(elev_mat) <- substr(colnames(elev_mat), start = 1, stop = 3)
colnames(elev_mat)[colnames(elev_mat) == 'ARB'] <- "ARU"
colnames(elev_mat)[colnames(elev_mat) == 'SPE'] <- "SAL"
rownames(elev_mat) <- colnames(elev_mat)
elev_mat <- makeSym(elev_mat)
elev_mat <- as.matrix(elev_mat[elev.nam , elev.nam])
row.labs <- paste0(elev.nam, " - ", elev[order(elev)], "m")


pheatmap(elev_mat, cluster_rows = FALSE, cluster_cols = FALSE, color = colorRampPalette(brewer.pal(n = 9, name =
  "YlOrRd"))(100), border_color = NA, labels_row = row.labs, fontsize_row = 15, fontsize_col = 15, filename = "C:/Users/Sophie/Michigan State University/Conner, Jeffrey - SophieAnalyses/Figures/ManuscriptFigs/supp/fst_heatmap_elevation_Dec2022_10k.png",height = 9, width = 9)

# cool! This went way easier than I thought it would. I should try and find a resource that uses diveRsity so I make sure I have something to cite and there isn't general backlash against this method...

##### Now with when Fst was calculated with population identifiers #####

sub_10k_pop <- read.csv("C:/Users/Sophie/Michigan State University/Conner, Jeffrey - SophieAnalyses/PopGen/fastDivPart_plinkFilter_sub_SNPs_Pop_Dec16_2022/pairwiseFst_WC.csv", header = TRUE)

## reorganize
rownames(sub_10k_pop) <- sub_10k_pop[,1]
sub_10k_pop <- sub_10k_pop[1:16,2:17]
sub_10k_pop <- lapply(sub_10k_pop, as.numeric)
sub_10k_pop <- as.matrix(as.data.frame(sub_10k_pop))
rownames(sub_10k_pop) <- colnames(sub_10k_pop)
# simplify for plotting
rownames(sub_10k_pop) <- substr(rownames(sub_10k_pop), start = 1, stop = 3)
colnames(sub_10k_pop) <- substr(colnames(sub_10k_pop), start = 1, stop = 3)

# get a mean and sd
stats_pop <- as.vector(sub_10k_pop)
sum_fst_pop <- c(mean(stats_pop, na.rm = TRUE), sd(stats_pop, na.rm = TRUE), min(stats_pop, na.rm = TRUE), max(stats_pop, na.rm = TRUE))
# mean = 0.478 sd = 0.17
# min = 0.115
# max = 0.868

sub_10k_pop <- makeSym(sub_10k_pop)
str(sub_10k_pop)

diag(sub_10k_pop) <- 0

str(sub_10k_pop)
fst_mat <- sub_10k_pop
substr(rownames(fst_mat),1,3)[substr(rownames(fst_mat),1,3) == 'ARB'] <- "ARU"
substr(colnames(fst_mat),1,3)[substr(colnames(fst_mat),1,3) == 'ARB'] <- "ARU"
substr(rownames(fst_mat),1,3)[substr(rownames(fst_mat),1,3) == 'SPE'] <- "SAL"
substr(colnames(fst_mat),1,3)[substr(colnames(fst_mat),1,3) == 'SPE'] <- "SAL"
save(fst_mat, file = "C:/Users/Sophie/Michigan State University/Conner, Jeffrey - SophieAnalyses/Fst_matrix_Dec2022_10k_pop.ROBJ")

pdf("C:/Users/Sophie/Michigan State University/Conner, Jeffrey - SophieAnalyses/Figures/Fst_heatmap_Dec2022_10k_pop.pdf", height = 9, width = 11)
heatmap(x = fst_mat, symm = TRUE,
        xlab = "Population",
        ylab = "Population",
        main = "Fst Heatmap", 
        keep.dendro = FALSE,
        #Rowv = NA,
        col= colorRampPalette(brewer.pal(8, "Oranges"))(255))
legend(x="bottomright", legend=c("0", "0.5", "1"),
     fill=colorRampPalette(brewer.pal(8, "Oranges"))(3))
dev.off()

## add in elevation information - run from above if only doing this, but can use elev lists from earlier
elev_mat <- as.data.frame(sub_10k_pop)
colnames(elev_mat)[colnames(elev_mat) == 'ARB'] <- "ARU"
colnames(elev_mat)[colnames(elev_mat) == 'SPE'] <- "SAL"
rownames(elev_mat) <- colnames(elev_mat)
elev_mat <- as.matrix(elev_mat[elev.nam , elev.nam])

pheatmap(elev_mat, cluster_rows = FALSE, cluster_cols = FALSE, color = colorRampPalette(brewer.pal(n = 9, name =
  "YlOrRd"))(100), border_color = NA, labels_row = row.labs, fontsize_row = 15, fontsize_col = 15, filename = "C:/Users/Sophie/Michigan State University/Conner, Jeffrey - SophieAnalyses/Figures/fst_heatmap_elevation_Dec2022_10k_pop.png",height = 9, width = 9)

```

## Without Centromeres
Repeat for the no centromere file

Note: code seems to be out of order here but I did the Fst calculation without the centromeres, and then visualized it so this code is reading in the no centromere results. I might need to search around on the HPCC to find that code.
```{r fst_heatmap2, eval = TRUE}
# clear everything
rm(list= ls())

##### visualizing fst matrices
library(RColorBrewer)
library(pheatmap)

## to fill in top part of matrix
makeSym <- function(m) {
  m[upper.tri(m)] <- t(m)[upper.tri(m)]
  return(m)
}

##### start with when Fst was calculated with individual identifiers #####

## read in the matrix
sub_10k <- read.csv("C:/Users/Sophie/Michigan State University/Conner, Jeffrey - SophieAnalyses/PopGen/fastDivPart_plinkFilter_sub_NoCent_SNPs_Jan5_2023/pairwiseFst_WC.csv", header = TRUE)

## reorganize
rownames(sub_10k) <- sub_10k[,1]
sub_10k <- sub_10k[1:16,2:17]
sub_10k <- lapply(sub_10k, as.numeric)
sub_10k <- as.matrix(as.data.frame(sub_10k))
rownames(sub_10k) <- colnames(sub_10k)
# simplify for plotting
rownames(sub_10k) <- substr(rownames(sub_10k), start = 1, stop = 3)
colnames(sub_10k) <- substr(colnames(sub_10k), start = 1, stop = 3)
str(sub_10k)

# get a mean and sd
stats <- as.vector(sub_10k)
sum_fst <- c(mean(stats, na.rm = TRUE), sd(stats, na.rm = TRUE), min(stats, na.rm = TRUE), max(stats, na.rm = TRUE))
# mean= 0.504, sd = 0.178
# min = 0.111
# max = 0.91

sub_10k <- makeSym(sub_10k)
str(sub_10k)

diag(sub_10k) <- 0

str(sub_10k)
fst_mat <- sub_10k
substr(rownames(fst_mat),1,3)[substr(rownames(fst_mat),1,3) == 'ARB'] <- "ARU"
substr(colnames(fst_mat),1,3)[substr(colnames(fst_mat),1,3) == 'ARB'] <- "ARU"
substr(rownames(fst_mat),1,3)[substr(rownames(fst_mat),1,3) == 'SPE'] <- "SAL"
substr(colnames(fst_mat),1,3)[substr(colnames(fst_mat),1,3) == 'SPE'] <- "SAL"
save(fst_mat, file = "C:/Users/Sophie/Michigan State University/Conner, Jeffrey - SophieAnalyses/Fst_matrix_NoCent_Dec2022_10k.ROBJ")

pdf("C:/Users/Sophie/Michigan State University/Conner, Jeffrey - SophieAnalyses/Figures/Fst_heatmap_NoCent_Dec2022_10k.pdf", height = 9, width = 11)
heatmap(x = fst_mat, symm = TRUE,
        xlab = "Population",
        ylab = "Population",
        main = "Fst Heatmap", 
        keep.dendro = FALSE,
        #Rowv = NA,
        col= colorRampPalette(brewer.pal(8, "Oranges"))(255))
legend(x="bottomright", legend=c("0", "0.5", "1"),
     fill=colorRampPalette(brewer.pal(8, "Oranges"))(3))
dev.off()

## add in elevation information
elev.nam <- c('RAB', 'PIN', 'SAL', 'HOR', 'ARU', 'BAR', 'COC', 'POB', 'BOS', 'MUR', 'VDM', 'ALE', 'BIS', 'PAL', 'VIE', 'PAN')
elev <- c(1229, 416, 441, 1444, 715, 519, 413, 836, 1585, 1706, 177, 665, 61, 303, 991, 1605)
elev_mat <- as.data.frame(sub_10k)
colnames(elev_mat)[colnames(elev_mat) == 'ARB'] <- "ARU"
colnames(elev_mat)[colnames(elev_mat) == 'SPE'] <- "SAL"
rownames(elev_mat) <- colnames(elev_mat)
elev_mat <- as.matrix(elev_mat[elev.nam , elev.nam])
row.labs <- paste0(elev.nam, " - ", elev[order(elev)], "m")

pheatmap(elev_mat, cluster_rows = FALSE, cluster_cols = FALSE, color = colorRampPalette(brewer.pal(n = 9, name =
  "YlOrRd"))(100), border_color = NA, labels_row = row.labs, fontsize_row = 15, fontsize_col = 15, filename = "C:/Users/Sophie/Michigan State University/Conner, Jeffrey - SophieAnalyses/Figures/ManuscriptFigs/supp/fst_heatmap_NoCent_elevation_Dec2022_10k.png",height = 9, width = 9)


##### Now with when Fst was calculated with population identifiers #####

sub_10k_pop <- read.csv("C:/Users/Sophie/Michigan State University/Conner, Jeffrey - SophieAnalyses/PopGen/fastDivPart_plinkFilter_sub_NoCent_SNPs_Pop_Jan6_2023/pairwiseFst_WC.csv", header = TRUE)

## reorganize
rownames(sub_10k_pop) <- sub_10k_pop[,1]
sub_10k_pop <- sub_10k_pop[1:16,2:17]
sub_10k_pop <- lapply(sub_10k_pop, as.numeric)
sub_10k_pop <- as.matrix(as.data.frame(sub_10k_pop))
rownames(sub_10k_pop) <- colnames(sub_10k_pop)

#change some row names
substr(rownames(sub_10k_pop),1,3)[substr(rownames(sub_10k_pop),1,3) == 'ARB'] <- "ARU"
substr(colnames(sub_10k_pop),1,3)[substr(colnames(sub_10k_pop),1,3) == 'ARB'] <- "ARU"
substr(rownames(sub_10k_pop),1,3)[substr(rownames(sub_10k_pop),1,3) == 'SPE'] <- "SAL"
substr(colnames(sub_10k_pop),1,3)[substr(colnames(sub_10k_pop),1,3) == 'SPE'] <- "SAL"

# get a mean and sd
stats_pop <- as.vector(sub_10k_pop)
sum_fst_pop <- c(mean(stats_pop, na.rm = TRUE), sd(stats_pop, na.rm = TRUE), min(stats_pop, na.rm = TRUE), max(stats_pop, na.rm = TRUE))
# mean = 0.506 sd = 0.178
# min = 0.119
# max = 0.914

sub_10k_pop <- makeSym(sub_10k_pop)
str(sub_10k_pop)

diag(sub_10k_pop) <- 0

str(sub_10k_pop)

fst_mat <- sub_10k_pop
save(fst_mat, file = "C:/Users/Sophie/Michigan State University/Conner, Jeffrey - SophieAnalyses/Fst_matrix_NoCent_Dec2022_10k_pop.ROBJ")

pdf("C:/Users/Sophie/Michigan State University/Conner, Jeffrey - SophieAnalyses/Figures/Fst_heatmap_NoCent_Dec2022_10k_pop.pdf", height = 9, width = 11)
heatmap(x = fst_mat, symm = TRUE,
        xlab = "Population",
        ylab = "Population",
        main = "Fst Heatmap", 
        keep.dendro = FALSE,
        #Rowv = NA,
        col= colorRampPalette(brewer.pal(8, "Oranges"))(255))
legend(x="bottomright", legend=c("0", "0.5", "1"),
     fill=colorRampPalette(brewer.pal(8, "Oranges"))(3))
dev.off()

## add in elevation information - run from above if only doing this, but can use elev lists from earlier
elev_mat <- as.data.frame(sub_10k_pop)
elev_mat <- as.matrix(elev_mat[elev.nam , elev.nam])

pheatmap(elev_mat, cluster_rows = FALSE, cluster_cols = FALSE, color = colorRampPalette(brewer.pal(n = 9, name =
  "YlOrRd"))(100), border_color = NA, labels_row = row.labs, fontsize_row = 15, fontsize_col = 15, filename = "C:/Users/Sophie/Michigan State University/Conner, Jeffrey - SophieAnalyses/Figures/fst_heatmap_NoCent_elevation_Dec2022_10k_pop.png",height = 9, width = 9)

```

Based on these results, there is not really a difference in if the centromere is removed or not for calculating Fst.Values without the centromere are a little bit higher. This might be because I am only choosing a small proportion of the snps to use to calculate Fst so it is unlikely I was choosing snps in the centromere region.

Also doesn't seem to matter if the calculation had the full ID names or just the Pop names, which is expected because I tell the diveRsity code that only the first three characters of the ID name determine the population