---
title: "08_PCA"
author: "Sophia Buysse"
date: "5/16/2022"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, eval = FALSE)
```

This uses the same input file as the GWAS.

# Do PCA

Doing it with plink is surprisingly easy!

```{bash}
module load PLINK/1.9b_4.1-x86_64

#all.filtered file. This is the file that is the input for the Fst, so it is variant and invariant sites that have been less stringently filtered
plink -bfile /mnt/scratch/buysseso/GVCF/all.filtered --threads 4 --pca --out /mnt/scratch/buysseso/PCA/plinkPCA_Var_invar

# plink filtering (see filter variants code)
plink -bfile /mnt/scratch/buysseso/GVCF/allsites_filtered_plinkTest --threads 4 --pca --out /mnt/scratch/buysseso/PCA/plinkPCA_plinkFilterTest

#plink filtering, no centromere
plink -bfile /mnt/scratch/buysseso/GVCF/allChrs_SNPs_NoCent --threads 4 --pca --out /mnt/scratch/buysseso/PCA/plinkPCA_noCent

### plink filtering, remove BOS individuals

# remove BOS individuals
# with cent
plink -bfile /mnt/scratch/buysseso/GVCF/allsites_filtered_plinkTest --threads 4 --remove /mnt/scratch/buysseso/PCA/Bos_indv.txt --make-bed --out /mnt/scratch/buysseso/GVCF/allsites_plink_noBOS
# output: 57 individuals remaining
# 1858706 variants remain
plink -bfile /mnt/scratch/buysseso/GVCF/allsites_plink_noBOS --threads 4 --pca --out /mnt/scratch/buysseso/PCA/plinkPCA_noBOS

# no cent and no BOS
plink -bfile /mnt/scratch/buysseso/GVCF/allChrs_SNPs_NoCent --threads 4 --remove /mnt/scratch/buysseso/PCA/Bos_indv.txt --make-bed --out /mnt/scratch/buysseso/GVCF/allChrs_noCent_noBOS
plink -bfile /mnt/scratch/buysseso/GVCF/allChrs_noCent_noBOS --threads 4 --pca --out /mnt/scratch/buysseso/PCA/plinkPCA_nocent_noBOS
```

Outputs!

For the all with variant and invariant sites:
[buysseso@dev-amd20 buysseso]$ plink -bfile /mnt/scratch/buysseso/GVCF/all.filtered --pca --out /mnt/scratch/buysseso/PCA/plinkPCA_Var_invar
PLINK v1.90b6.6 64-bit (10 Oct 2018)           www.cog-genomics.org/plink/1.9/
(C) 2005-2018 Shaun Purcell, Christopher Chang   GNU General Public License v3
Logging to /mnt/scratch/buysseso/PCA/plinkPCA_Var_invar.log.
Options in effect:
  --bfile /mnt/scratch/buysseso/GVCF/all.filtered
  --out /mnt/scratch/buysseso/PCA/plinkPCA_Var_invar
  --pca

1031524 MB RAM detected; reserving 515762 MB for main workspace.
99202614 variants loaded from .bim file.
61 people (0 males, 0 females, 61 ambiguous) loaded from .fam.
Ambiguous sex IDs written to /mnt/scratch/buysseso/PCA/plinkPCA_Var_invar.nosex
.
Using up to 127 threads (change this with --threads).
Before main variant filters, 61 founders and 0 nonfounders present.
Calculating allele frequencies... done.
Total genotyping rate is 0.959406.
99202614 variants and 61 people pass filters and QC.
Note: No phenotypes present.
Relationship matrix calculation complete.
--pca: Results saved to /mnt/scratch/buysseso/PCA/plinkPCA_Var_invar.eigenval
and /mnt/scratch/buysseso/PCA/plinkPCA_Var_invar.eigenvec .

For plink filtering allsites
[buysseso@dev-amd20 GVCF]$ plink -bfile /mnt/scratch/buysseso/GVCF/allsites_filtered_plinkTest --threads 4 --pca --out /mnt/scratch/buysseso/PCA/plinkPCA_plinkFilterTest
PLINK v1.90b6.6 64-bit (10 Oct 2018)           www.cog-genomics.org/plink/1.9/
(C) 2005-2018 Shaun Purcell, Christopher Chang   GNU General Public License v3
Logging to /mnt/scratch/buysseso/PCA/plinkPCA_plinkFilterTest.log.
Options in effect:
  --bfile /mnt/scratch/buysseso/GVCF/allsites_filtered_plinkTest
  --out /mnt/scratch/buysseso/PCA/plinkPCA_plinkFilterTest
  --pca
  --threads 4

1031524 MB RAM detected; reserving 515762 MB for main workspace.
1858706 variants loaded from .bim file.
61 people (0 males, 0 females, 61 ambiguous) loaded from .fam.
Ambiguous sex IDs written to
/mnt/scratch/buysseso/PCA/plinkPCA_plinkFilterTest.nosex .
Using up to 4 threads (change this with --threads).
Warning: This run includes BLAS/LAPACK linear algebra operations which
currently disregard the --threads limit.  If this is problematic, you may want
to recompile against single-threaded BLAS/LAPACK.
Before main variant filters, 61 founders and 0 nonfounders present.
Calculating allele frequencies... done.
Total genotyping rate is 0.932702.
1858706 variants and 61 people pass filters and QC.
Note: No phenotypes present.
Relationship matrix calculation complete.
--pca: Results saved to and eigenvectors]
/mnt/scratch/buysseso/PCA/plinkPCA_plinkFilterTest.eigenval and
/mnt/scratch/buysseso/PCA/plinkPCA_plinkFilterTest.eigenvec .

removing BOS from nocent:
plink -bfile /mnt/scratch/buysseso/GVCF/allChrs_SNPs_NoCent --threads 4 --remove /mnt/scratch/buysseso/PCA/Bos_indv.txt --make-bed --out /mnt/scratch/buysseso/GVCF/allChrs_noCent_noBOS
PLINK v1.90b6.6 64-bit (10 Oct 2018)           www.cog-genomics.org/plink/1.9/
(C) 2005-2018 Shaun Purcell, Christopher Chang   GNU General Public License v3
Logging to /mnt/scratch/buysseso/GVCF/allChrs_noCent_noBOS.log.
Options in effect:
  --bfile /mnt/scratch/buysseso/GVCF/allChrs_SNPs_NoCent
  --make-bed
  --out /mnt/scratch/buysseso/GVCF/allChrs_noCent_noBOS
  --remove /mnt/scratch/buysseso/PCA/Bos_indv.txt
  --threads 4

1031524 MB RAM detected; reserving 515762 MB for main workspace.
1507328 variants loaded from .bim file.
61 people (0 males, 0 females, 61 ambiguous) loaded from .fam.
Ambiguous sex IDs written to
/mnt/scratch/buysseso/GVCF/allChrs_noCent_noBOS.nosex .
--remove: 57 people remaining.
Using 1 thread (no multithreaded calculations invoked).
Before main variant filters, 57 founders and 0 nonfounders present.
Calculating allele frequencies... done.
Total genotyping rate in remaining samples is 0.933517.
1507328 variants and 57 people pass filters and QC.
Note: No phenotypes present.
--make-bed to /mnt/scratch/buysseso/GVCF/allChrs_noCent_noBOS.bed +
/mnt/scratch/buysseso/GVCF/allChrs_noCent_noBOS.bim +
/mnt/scratch/buysseso/GVCF/allChrs_noCent_noBOS.fam ... done.

noBOS pCA output:
1031524 MB RAM detected; reserving 515762 MB for main workspace.
1858706 variants loaded from .bim file.
57 people (0 males, 0 females, 57 ambiguous) loaded from .fam.
Ambiguous sex IDs written to /mnt/scratch/buysseso/PCA/plinkPCA_noBOS.nosex .
Using up to 4 threads (change this with --threads).
Warning: This run includes BLAS/LAPACK linear algebra operations which
currently disregard the --threads limit.  If this is problematic, you may want
to recompile against single-threaded BLAS/LAPACK.
Before main variant filters, 57 founders and 0 nonfounders present.
Calculating allele frequencies... done.
Total genotyping rate is 0.930465.
1858706 variants and 57 people pass filters and QC.
Note: No phenotypes present.
Relationship matrix calculation complete.
--pca: Results saved to /mnt/scratch/buysseso/PCA/plinkPCA_noBOS.eigenval and
/mnt/scratch/buysseso/PCA/plinkPCA_noBOS.eigenvec .

no BOS no cent PCA output:
plink -bfile /mnt/scratch/buysseso/GVCF/allChrs_noCent_noBOS --threads 4 --pca --out /mnt/scratch/buysseso/PCA/plinkPCA_nocent_noBOS
PLINK v1.90b6.6 64-bit (10 Oct 2018)           www.cog-genomics.org/plink/1.9/
(C) 2005-2018 Shaun Purcell, Christopher Chang   GNU General Public License v3
Logging to /mnt/scratch/buysseso/PCA/plinkPCA_nocent_noBOS.log.
Options in effect:
  --bfile /mnt/scratch/buysseso/GVCF/allChrs_noCent_noBOS
  --out /mnt/scratch/buysseso/PCA/plinkPCA_nocent_noBOS
  --pca
  --threads 4

1031524 MB RAM detected; reserving 515762 MB for main workspace.
1507328 variants loaded from .bim file.
57 people (0 males, 0 females, 57 ambiguous) loaded from .fam.
Ambiguous sex IDs written to
/mnt/scratch/buysseso/PCA/plinkPCA_nocent_noBOS.nosex .
Using up to 4 threads (change this with --threads).
Warning: This run includes BLAS/LAPACK linear algebra operations which
currently disregard the --threads limit.  If this is problematic, you may want
to recompile against single-threaded BLAS/LAPACK.
Before main variant filters, 57 founders and 0 nonfounders present.
Calculating allele frequencies... done.
Total genotyping rate is 0.933517.
1507328 variants and 57 people pass filters and QC.
Note: No phenotypes present.
Relationship matrix calculation complete.
--pca: Results saved to and eigenvectors]
/mnt/scratch/buysseso/PCA/plinkPCA_nocent_noBOS.eigenval and
/mnt/scratch/buysseso/PCA/plinkPCA_nocent_noBOS.eigenvec .


Plink filtering, no cent
1031524 MB RAM detected; reserving 515762 MB for main workspace.
1507328 variants loaded from .bim file.
61 people (0 males, 0 females, 61 ambiguous) loaded from .fam.
Ambiguous sex IDs written to /mnt/scratch/buysseso/PCA/plinkPCA_noCent.nosex .
Using up to 4 threads (change this with --threads).
Warning: This run includes BLAS/LAPACK linear algebra operations which
currently disregard the --threads limit.  If this is problematic, you may want
to recompile against single-threaded BLAS/LAPACK.
Before main variant filters, 61 founders and 0 nonfounders present.
Calculating allele frequencies... done.
Total genotyping rate is 0.936191.
1507328 variants and 61 people pass filters and QC.
Note: No phenotypes present.
Relationship matrix calculation complete.
--pca: Results saved to /mnt/scratch/buysseso/PCA/plinkPCA_noCent.eigenval and
/mnt/scratch/buysseso/PCA/plinkPCA_noCent.eigenvec .


# Now visualize the results with R
### Prep Datasets ###

manually moved important data file to my local machine to make higher quality figures.

```{r}
library(dplyr)
library(ggplot2)

# read in data
# plink filter, all
eigenValues_test <- read.delim("~/R_script/StamenLossPipeline/data/plinkPCA_plinkFilterTest.eigenval", sep = " ", header = F)
eigenVectors_test <- read.delim("~/R_script/StamenLossPipeline/data/plinkPCA_plinkFilterTest.eigenvec", sep = " ", header = F)
# each row is a genotype, each column is a PC.
colnames(eigenVectors_test) <- c("Line", "SeqID", paste0("PC", 1:(ncol(eigenVectors_test)-2)))

# plink filter, no BOS
eigenValues_noBOS <- read.delim("~/R_script/StamenLossPipeline/data/plinkPCA_noBOS.eigenval", sep = " ", header = F)
eigenVectors_noBOS <- read.delim("~/R_script/StamenLossPipeline/data/plinkPCA_noBOS.eigenvec", sep = " ", header = F)
# each row is a genotype, each column is a PC.
colnames(eigenVectors_noBOS) <- c("Line", "SeqID", paste0("PC", 1:(ncol(eigenVectors_noBOS)-2)))

# plink filter, no centromere
eigenValues_noCent <- read.delim("~/R_script/StamenLossPipeline/data/plinkPCA_noCent.eigenval", sep = " ", header = F)
eigenVectors_noCent <- read.delim("~/R_script/StamenLossPipeline/data/plinkPCA_noCent.eigenvec", sep = " ", header = F)
# each row is a genotype, each column is a PC.
colnames(eigenVectors_noCent) <- c("Line", "SeqID", paste0("PC", 1:(ncol(eigenVectors_noCent)-2)))

# plink filter, no centromere, no BOS
eigenValues_noCent_noBOS <- read.delim("~/R_script/StamenLossPipeline/data/plinkPCA_nocent_noBOS.eigenval", sep = " ", header = F)
eigenVectors_noCent_noBOS <- read.delim("~/R_script/StamenLossPipeline/data/plinkPCA_nocent_noBOS.eigenvec", sep = " ", header = F)
# each row is a genotype, each column is a PC.
colnames(eigenVectors_noCent_noBOS) <- c("Line", "SeqID", paste0("PC", 1:(ncol(eigenVectors_noCent_noBOS)-2)))

# calculate the PVE. Remember that I used 20 PCs, so this PVE is out of 20 PCs
pve_test <- round((eigenValues_test / (sum(eigenValues_test))*100), 2)
pve_test$PC <- c(1:20)

pve_noBOS <- round((eigenValues_noBOS / (sum(eigenValues_noBOS))*100), 2)
pve_noBOS$PC <- c(1:20)
# definite drop after first four, 1 is 13, 2 is 10, 3 is 8.3, 4 is 7.` and the rest scale from 6 down to 2.25

pve_noCent <- round((eigenValues_noCent / (sum(eigenValues_noCent))*100), 2)
pve_noCent$PC <- c(1:20)
# matches the NO BOS results almost exactly? strange...

pve_noCent_noBOS <- round((eigenValues_noCent_noBOS / (sum(eigenValues_noCent_noBOS))*100), 2)
pve_noCent_noBOS$PC <- c(1:20)

# add pop column
eigenVectors_test$Pop <- toupper(substr(eigenVectors_test$Line, 1, 3))
eigenVectors_noBOS$Pop <- toupper(substr(eigenVectors_noBOS$Line, 1, 3))
eigenVectors_noCent$Pop <- toupper(substr(eigenVectors_noCent$Line, 1, 3))
eigenVectors_noCent_noBOS$Pop <- toupper(substr(eigenVectors_noCent_noBOS$Line, 1, 3))

# change ARB -> ARU and SPE -> SAL
eigenVectors_test[eigenVectors_test$Pop == "ARB", "Pop"] <- "ARU"
eigenVectors_test[eigenVectors_test$Pop == "SPE", "Pop"] <- "SAL"

eigenVectors_noBOS[eigenVectors_noBOS$Pop == "ARB", "Pop"] <- "ARU"
eigenVectors_noBOS[eigenVectors_noBOS$Pop == "SPE", "Pop"] <- "SAL"

eigenVectors_noCent[eigenVectors_noCent$Pop == "ARB", "Pop"] <- "ARU"
eigenVectors_noCent[eigenVectors_noCent$Pop == "SPE", "Pop"] <- "SAL"

eigenVectors_noCent_noBOS[eigenVectors_noCent_noBOS$Pop == "ARB", "Pop"] <- "ARU"
eigenVectors_noCent_noBOS[eigenVectors_noCent_noBOS$Pop == "SPE", "Pop"] <- "SAL"

# load the meta data
metadata <- read.csv("~/SpanishMasterDataset_PopMetaData.csv", header = TRUE)

# clean up a little
metadata <- metadata[1:16,c("Population", "PopCode", "Elev_m", "Lat_DecDeg", "Lon_DecDeg")]

# merge 
eigenVectors_test$elev <- metadata$Elev_m[match(eigenVectors_test$Pop, metadata$PopCode)]
# then match with row names to add metadata column
# do this for lat and long now too
eigenVectors_test$lat <- metadata$Lat_DecDeg[match(eigenVectors_test$Pop, metadata$PopCode)]
eigenVectors_test$lon <- metadata$Lon_DecDeg[match(eigenVectors_test$Pop, metadata$PopCode)]

eigenVectors_noBOS$elev <- metadata$Elev_m[match(eigenVectors_noBOS$Pop, metadata$PopCode)]
# then match with row names to add metadata column
# do this for lat and long now too
eigenVectors_noBOS$lat <- metadata$Lat_DecDeg[match(eigenVectors_noBOS$Pop, metadata$PopCode)]
eigenVectors_noBOS$lon <- metadata$Lon_DecDeg[match(eigenVectors_noBOS$Pop, metadata$PopCode)]

eigenVectors_noCent$elev <- metadata$Elev_m[match(eigenVectors_noCent$Pop, metadata$PopCode)]
# then match with row names to add metadata column
# do this for lat and long now too
eigenVectors_noCent$lat <- metadata$Lat_DecDeg[match(eigenVectors_noCent$Pop, metadata$PopCode)]
eigenVectors_noCent$lon <- metadata$Lon_DecDeg[match(eigenVectors_noCent$Pop, metadata$PopCode)]

eigenVectors_noCent_noBOS$elev <- metadata$Elev_m[match(eigenVectors_noCent_noBOS$Pop, metadata$PopCode)]
# then match with row names to add metadata column
# do this for lat and long now too
eigenVectors_noCent_noBOS$lat <- metadata$Lat_DecDeg[match(eigenVectors_noCent_noBOS$Pop, metadata$PopCode)]
eigenVectors_noCent_noBOS$lon <- metadata$Lon_DecDeg[match(eigenVectors_noCent_noBOS$Pop, metadata$PopCode)]

# order by elevation and make labels column

eigenVectors_test <- eigenVectors_test[order(eigenVectors_test$elev),]
eigenVectors_test$label <- paste0(eigenVectors_test$Pop, " - ", eigenVectors_test$elev, "m")
#eigenVectors_test$Pop <- as.factor(eigenVectors_test$Pop)
eigenVectors_test$Pop <- factor(eigenVectors_test$Pop, levels = unique(eigenVectors_test$Pop[order(eigenVectors_test$elev)]))
str(eigenVectors_test)

eigenVectors_noBOS <- eigenVectors_noBOS[order(eigenVectors_noBOS$elev),]
eigenVectors_noBOS$label <- paste0(eigenVectors_noBOS$Pop, " - ", eigenVectors_noBOS$elev, "m")
eigenVectors_noBOS$Pop <- factor(eigenVectors_noBOS$Pop, levels = unique(eigenVectors_noBOS$Pop[order(eigenVectors_noBOS$elev)]))

eigenVectors_noCent <- eigenVectors_noCent[order(eigenVectors_noCent$elev),]
eigenVectors_noCent$label <- paste0(eigenVectors_noCent$Pop, " - ", eigenVectors_noCent$elev, "m")
eigenVectors_noCent$Pop <- factor(eigenVectors_noCent$Pop, levels = unique(eigenVectors_noCent$Pop[order(eigenVectors_noCent$elev)]))

eigenVectors_noCent_noBOS <- eigenVectors_noCent_noBOS[order(eigenVectors_noCent_noBOS$elev),]
eigenVectors_noCent_noBOS$label <- paste0(eigenVectors_noCent_noBOS$Pop, " - ", eigenVectors_noCent_noBOS$elev, "m")
eigenVectors_noCent_noBOS$Pop <- factor(eigenVectors_noCent_noBOS$Pop, levels = unique(eigenVectors_noCent_noBOS$Pop[order(eigenVectors_noCent_noBOS$elev)]))
```

### Make Plots ###
Plink filtering, includes centromere
```{r}
# set.seed for jitter
set.seed(656)
##### plink filtering test #####
# barplot of PVE 
ggplot(pve_test)+ geom_bar(aes(y = V1,x = PC), stat = "identity")
# this looks more similar to the all.filtered output with higher values than the vcftools filtered file.

# TOPO graphs
# alpha (semi-transparency) not supported on HPCC (got a not supported on this device error)

# topo fill PC1/PC2
plot_3a <- ggplot(data = eigenVectors_test) +
  geom_jitter(mapping = aes(x = -(PC1), y = PC2, shape = Pop, fill = elev ), col = "black", size = 1.3, stroke = 0.5, width = 0.025, height=0.025, show.legend = FALSE ) +
  geom_hline(yintercept = 0, linetype="dotted") +
  geom_vline(xintercept = 0, linetype="dotted") +
  labs(x = paste0("PC 1 (",sprintf(pve_test[1,1], fmt='%#.1f'),"%)"),
      y = paste0("PC 2 (",sprintf(pve_test[2,1], fmt='%#.1f'),"%)")) +
  scale_shape_manual(name = "Population",
                     labels = unique(eigenVectors_test$label),
                     values = rep(c(22, 21, 24, 23, 25), times = 4))+
  scale_fill_gradientn(name = "Elevation", colours = topo.colors(16))+
  theme_classic()+
  theme(
    legend.title = element_text(color = "black", size = 12),
    legend.text = element_text(color = "black", size = 12),
    axis.title.y = element_text(color = "black", size = 10),
    axis.title.x = element_text(color = "black", size = 10),
    axis.text = element_text(color = "black", size = 10),
    legend.spacing.y = unit(0.03, "cm"))

annotate_figure(plot_3a,
                fig.lab = "A", fig.lab.face = "bold")

ggsave(filename = "plink_PC1PC2_PlinkFiltering_topo_fill.png", path = "~/Figures/ManuscriptFigs", height = 3, width = 3, device = "png", dpi = 700)

## topo fill PC3/PC4
plot_s4a <- ggplot(data = eigenVectors_test) +
  geom_jitter(mapping = aes(x = PC3, y = PC4, shape = Pop, fill = elev ), col = "black", size = 1.3, stroke = 0.5, height = 0.025, width=0.025, show.legend = FALSE ) + # was size = 3, stroke = 1.25 alpha = 0.7,
  geom_hline(yintercept = 0, linetype="dotted") +
  geom_vline(xintercept = 0, linetype="dotted") +
  labs(x = paste0("PC 3 (",sprintf(pve_test[3,1], fmt='%#.1f'),"%)"),
      y = paste0("PC 4 (",sprintf(pve_test[4,1], fmt='%#.1f'),"%)")) +
  scale_shape_manual(name = "Population",
                     labels = unique(eigenVectors_test$label),
                     values = rep(c(22, 21, 24, 23, 25), times = 4))+
  scale_fill_gradientn(name = "Elevation", colours = topo.colors(16))+
  theme_classic()+
  theme(
    legend.title = element_text(color = "black", size = 12),
    legend.text = element_text(color = "black", size = 12),
    axis.title = element_text(color = "black", size = 10),
    axis.text = element_text(color = "black", size = 10),
    legend.spacing.y = unit(0.03, "cm"))

annotate_figure(plot_s4a,
                fig.lab = "A", fig.lab.face = "bold")
ggsave(filename = "plink_PC3PC4_PlinkFiltering_topo_fill.png", path = "~/Figures/ManuscriptFigs/supp", height = 3, width = 3, device = "png", dpi = 700)

```

no BOS, includes the centromere
```{r}
##### plink, no BOS #####
# barplot of PVE 
ggplot(pve_noBOS)+ geom_bar(aes(y = V1,x = PC), stat = "identity")
# looks decent, don't really know what to note.

#topo plots

## topo fill PC1/PC2
plot_3b <- ggplot(data = eigenVectors_noBOS) +
  geom_jitter(mapping = aes(x = -PC1, y = PC2, shape = Pop, fill = elev ), col = "black", size = 1.3, stroke = 0.5, height = 0.025, width = 0.025, show.legend = FALSE ) +
  geom_hline(yintercept = 0, linetype="dotted") +
  geom_vline(xintercept = 0, linetype="dotted") +
  labs(x = paste0("PC 1 (",sprintf(pve_noBOS[1,1], fmt='%#.1f'),"%)"),
      y = paste0("PC 2 (",sprintf(pve_noBOS[2,1], fmt='%#.1f'),"%)")) +
  scale_shape_manual(name = "Population",
                     labels = unique(eigenVectors_noBOS$label),
                     values = c(22,21,24,23,25,22,21,24,25,22,21,24,23,25,22,21,24,23,25))+
  scale_fill_gradientn(name = "Elevation", colours = topo.colors(16))+
  theme_classic()+
  theme(
    legend.title = element_text(color = "black", size = 12),
    legend.text = element_text(color = "black", size = 12),
    axis.title = element_text(color = "black", size = 10),
    axis.text = element_text(color = "black", size = 10),
    legend.spacing.y = unit(0.03, "cm"))
annotate_figure(plot_3b,
                fig.lab = "B", fig.lab.face = "bold")
ggsave(filename = "plink_PC1PC2_PlinkFiltering_noBOS_topo_fill.png", path = "~/Figures/ManuscriptFigs", height = 3, width = 3, device = "png", dpi = 700)

## topo fill PC3/PC4
plot_s4b <- ggplot(data = eigenVectors_noBOS) +
  geom_jitter(mapping = aes(x = PC3, y = PC4, shape = Pop, fill = elev ), col = "black", size = 1.3, stroke = 0.5, height = 0.025, width=0.025, show.legend = FALSE ) +
  geom_hline(yintercept = 0, linetype="dotted") +
  geom_vline(xintercept = 0, linetype="dotted") +
  labs(x = paste0("PC 3 (",sprintf(pve_noBOS[3,1], fmt='%#.1f'),"%)"),
      y = paste0("PC 4 (",sprintf(pve_noBOS[4,1], fmt='%#.1f'),"%)")) +
  scale_shape_manual(name = "Population",
                     labels = unique(eigenVectors_noBOS$label),
                     values = c(22,21,24,23,25,22,21,24,25,22,21,24,23,25,22,21,24,23,25))+
  scale_fill_gradientn(name = "Elevation", colours = topo.colors(16))+
  theme_classic()+
  theme(
    legend.title = element_text(color = "black", size = 12),
    legend.text = element_text(color = "black", size = 12),
    axis.title = element_text(color = "black", size = 10),
    axis.text = element_text(color = "black", size = 10),
    legend.spacing.y = unit(0.03, "cm"))
annotate_figure(plot_s4b,
                fig.lab = "B", fig.lab.face = "bold")
ggsave(filename = "plink_PC3PC4_PlinkFiltering_noBOS_topo_fill.png", path = "~/Figures/ManuscriptFigs/supp", height = 3, width = 3, device = "png", dpi = 700)

```

No Centromere
```{r}
##### plink filtering, no centromere #####
# barplot of PVE 
ggplot(pve_noCent)+ geom_bar(aes(y = V1,x = PC), stat = "identity")
# nothing really notable.

# topo plots
## topo fill PC1/PC2
plot_s5a <- ggplot(data = eigenVectors_noCent) +
  geom_jitter(mapping = aes(x = PC1, y = PC2, shape = Pop, fill = elev ), col = "black", size = 1.3, stroke = 0.5, height = 0.025, width = 0.025, show.legend = FALSE ) +
  geom_hline(yintercept = 0, linetype="dotted") +
  geom_vline(xintercept = 0, linetype="dotted") +
  labs(x = paste0("PC 1 (",sprintf(pve_noCent[1,1], fmt='%#.f'),"%)"),
      y = paste0("PC 2 (",sprintf(pve_noCent[2,1], fmt='%#.f'),"%)")) +
  scale_shape_manual(name = "Population",
                     labels = unique(eigenVectors_noCent$label),
                     values = rep(c(22, 21, 24, 23, 25), times = 4))+
  scale_fill_gradientn(name = "Elevation", colours = topo.colors(16))+
  theme_classic()+
  theme(
    legend.title = element_text(color = "black", size = 12),
    legend.text = element_text(color = "black", size = 12),
    axis.title = element_text(color = "black", size = 10),
    axis.text = element_text(color = "black", size = 10),
    legend.spacing.y = unit(0.03, "cm"))
annotate_figure(plot_s5a,
                fig.lab = "A", fig.lab.face = "bold")
ggsave(filename = "plink_PC1PC2_PlinkFiltering_noCent_topo_fill.png", path = "~/Figures/ManuscriptFigs/supp", height = 3, width = 3, device = "png", dpi = 700)

## topo fill PC1/PC2 labelled points to check out ALe-4 and compare to heterozygosity
ggplot(data = eigenVectors_noCent) +
  geom_point(mapping = aes(x = PC1, y = PC2, shape = Pop, fill = elev ), col = "black", size = 4, stroke = 1, show.legend = FALSE ) + # was size = 3, stroke = 1.25 alpha = 0.7,
  #geom_text(data = eigenVectors_noCent[eigenVectors_noCent$SeqID == "Ale-4", ], aes(x = PC1, y = PC2, label = SeqID), nudge_x = 0, nudge_y = +0.02, col = "black")+
  geom_text(aes(x = PC1, y = PC2, label = SeqID), nudge_x = 0, nudge_y = +0.02, col = "black")+
  geom_hline(yintercept = 0, linetype="dotted") +
  geom_vline(xintercept = 0, linetype="dotted") +
  labs(title = "PCA of Pyrenees Pops: Plink Filtering, No Centromere",
      x = paste0("PC 1 (",sprintf(pve_noCent[1,1], fmt='%#.f'),"%)"),
      y = paste0("PC 2 (",sprintf(pve_noCent[2,1], fmt='%#.f'),"%)")) +
  scale_shape_manual(name = "Population",
                     labels = unique(eigenVectors_noCent$label),
                     values = rep(c(22, 21, 24, 23, 25), times = 4))+
  scale_fill_gradientn(name = "Elevation", colours = topo.colors(16))+
  theme_classic()+
  theme(
    legend.title = element_text(color = "black", size = 20),
    legend.text = element_text(color = "black", size = 20),
    axis.title = element_text(color = "black", size = 20),
    axis.text = element_text(color = "black", size = 20),
    legend.spacing.y = unit(0.03, "cm"))

ggsave(filename = "plink_PC1PC2_PlinkFiltering_noCent_topo_fill_labels.png", path = "~/Figures", height = 7, width = 7, device = "png", dpi = 500)

## topo fill PC3/PC4
plot_s5b <- ggplot(data = eigenVectors_noCent) +
  geom_jitter(mapping = aes(x = PC3, y = PC4, shape = Pop, fill = elev ), col = "black", size = 1.3, stroke = 0.5, height = 0.025, width = 0.025, show.legend = FALSE ) +
  #geom_text(aes(x = PC3, y = PC4, label = SeqID), nudge_x = 0, nudge_y = +0.02, col = "black")+
  geom_hline(yintercept = 0, linetype="dotted") +
  geom_vline(xintercept = 0, linetype="dotted") +
  labs(x = paste0("PC 3 (",sprintf(pve_noCent[3,1], fmt='%#.1f')," %)"),
      y = paste0("PC 4 (",sprintf(pve_noCent[4,1], fmt='%#.1f')," %)")) +
  scale_shape_manual(name = "Population",
                     labels = unique(eigenVectors_noCent$label),
                     values = rep(c(22, 21, 24, 23, 25), times = 4))+
  scale_fill_gradientn(name = "Elevation", colours = topo.colors(16))+
  theme_classic()+
  theme(
    legend.title = element_text(color = "black", size = 12),
    legend.text = element_text(color = "black", size = 12),
    axis.title = element_text(color = "black", size = 10),
    axis.text = element_text(color = "black", size = 10),
    legend.spacing.y = unit(0.03, "cm"))
annotate_figure(plot_s5b,
                fig.lab = "B", fig.lab.face = "bold")
ggsave(filename = "plink_PC3PC4_PlinkFiltering_noCent_topo_fill.png", path = "~/Figures/ManuscriptFigs/supp", height = 3, width = 3, device = "png", dpi = 700)

```

No Centromere, no BOS
```{r}
##### plink filtering, no centromere, no BOS #####

## topo fill PC1/PC2
plot_s5c <- ggplot(data = eigenVectors_noCent_noBOS) +
  geom_jitter(mapping = aes(x = PC1, y = PC2, shape = Pop, fill = elev ), col = "black", size = 1.3, stroke = 0.5, height = 0.025, width = 0.025, show.legend = FALSE ) + # was size = 3, stroke = 1.25 alpha = 0.7,
  geom_hline(yintercept = 0, linetype="dotted") +
  geom_vline(xintercept = 0, linetype="dotted") +
  labs(x = paste0("PC 1 (",sprintf(pve_noCent_noBOS[1,1], fmt='%#.1f'),"%)"),
      y = paste0("PC 2 (",sprintf(pve_noCent_noBOS[2,1], fmt='%#.1f'),"%)")) +
  scale_shape_manual(name = "Population",
                     labels = unique(eigenVectors_noCent_noBOS$label),
                     values = c(22,21,24,23,25,22,21,24,25,22,21,24,23,25,22,21,24,23,25))+
  scale_fill_gradientn(name = "Elevation", colours = topo.colors(16))+
  theme_classic()+
  theme(
    legend.title = element_text(color = "black", size = 12),
    legend.text = element_text(color = "black", size = 12),
    axis.title = element_text(color = "black", size = 10),
    axis.text = element_text(color = "black", size = 10),
    legend.spacing.y = unit(0.03, "cm"))
annotate_figure(plot_s5c,
                fig.lab = "C", fig.lab.face = "bold")
ggsave(filename = "plink_PC1PC2_PlinkFiltering_noCent_noBOS_topo_fill.png", path = "~/Figures/ManuscriptFigs/supp", height = 3, width = 3, device = "png", dpi = 700)

## topo fill PC3/PC4
plot_s5d <- ggplot(data = eigenVectors_noCent_noBOS) +
  geom_jitter(mapping = aes(x = PC3, y = PC4, shape = Pop, fill = elev ), col = "black", size = 1.3, stroke = 0.5, height = 0.025, width = 0.025, show.legend = FALSE ) + # was size = 3, stroke = 1.25 alpha = 0.7,
  geom_hline(yintercept = 0, linetype="dotted") +
  geom_vline(xintercept = 0, linetype="dotted") +
  labs(x = paste0("PC 3 (",sprintf(pve_noCent_noBOS[3,1], fmt='%#.1f'),"%)"),
      y = paste0("PC 4 (",sprintf(pve_noCent_noBOS[4,1], fmt='%#.1f'),"%)")) +
  scale_shape_manual(name = "Population",
                     labels = unique(eigenVectors_noCent_noBOS$label),
                     values = c(22,21,24,23,25,22,21,24,25,22,21,24,23,25,22,21,24,23,25))+
  scale_fill_gradientn(name = "Elevation", colours = topo.colors(16))+
  theme_classic()+
  theme(
    legend.title = element_text(color = "black", size = 12),
    legend.text = element_text(color = "black", size = 12),
    axis.title = element_text(color = "black", size = 10),
    axis.text = element_text(color = "black", size = 10),
    legend.spacing.y = unit(0.03, "cm"))
annotate_figure(plot_s5d,
                fig.lab = "D", fig.lab.face = "bold")
ggsave(filename = "plink_PC3PC4_PlinkFiltering_noCent_noBOS_topo_fill.png", path = "~/Figures/ManuscriptFigs/supp", height = 3, width = 3, device = "png", dpi = 700)
```

### Regressions with Geo Vars ###
Check for correlations between PCs and geographic variables. Assumes you start this code with dplyr and the merged eigenVectors and metadata file that was used to make graphs.

```{r}
# let's make a subset dataframe of the variables we are interested in
cor_test <- eigenVectors_test[,c("PC1", "PC2", "PC3", "PC4", "elev", "lat", "lon", "Pop")]

cor_noBOS <- eigenVectors_noBOS[,c("PC1", "PC2", "PC3", "PC4", "elev", "lat", "lon", "Pop")]

cor_noCent <- eigenVectors_noCent[,c("PC1", "PC2", "PC3", "PC4", "elev", "lat", "lon", "Pop")]

cor_noCent_noBOS <- eigenVectors_noCent_noBOS[,c("PC1", "PC2", "PC3", "PC4", "elev", "lat", "lon", "Pop")]

# then, I want a summary table with mean PC values per population

summary_test <- cor_test %>%
  group_by(Pop) %>%
  summarize("elev" = mean(elev), "lat" = mean(lat), "lon" = mean(lon), "PC1" = mean(PC1), "PC2" = mean(PC2), "PC3" = mean(PC3), "PC4" = mean(PC4))
round(cor(summary_test[ ,2:8]), 2)
# lat and long are highly correlated with PC2 but I don't know what that might biologically mean.

# summarize without BOS
summary_noBOS <- cor_noBOS %>%
  group_by(Pop) %>%
  summarize("elev" = mean(elev), "lat" = mean(lat), "lon" = mean(lon), "PC1" = mean(PC1), "PC2" = mean(PC2), "PC3" = mean(PC3), "PC4" = mean(PC4))
round(cor(summary_noBOS[ ,2:8]), 2)

# initial correlations

cor.test(summary_test$elev, summary_test$PC1)
# p = 0.0009217, cor = -0.745
cor.test(summary_test$elev, summary_test$PC2)
# p = 0.03083, cor = -0.5400
cor.test(summary_test$elev, summary_test$PC3)
# cor = -0.17, p = 0.5244
cor.test(summary_test$elev, summary_test$PC4)
# cor = -0.08, p = 0.7425

cor.test(summary_test$elev, summary_test$lat)
# cor = 0.58, p = 0.017
cor.test(summary_test$elev, summary_test$lon)
# cor = -0.83, p = 5.76e-5

# no BOS correlations
cor.test(summary_noBOS$elev, summary_noBOS$PC1)
# p = 1.148e-6, cor = -0.9205671
cor.test(summary_noBOS$elev, summary_noBOS$PC2)
# p = 0.551, cor = -0.1673805
cor.test(summary_noBOS$elev, summary_noBOS$PC3)
# cor = 0.1110533, p = 0.6936
cor.test(summary_noBOS$elev, summary_noBOS$PC4)
# cor = 0.143123, p = 0.6109

# then multiple regression of each PC on all three variables

PC1_geo <- lm(PC1 ~ elev + lat + lon, data = summary_test)
summary(PC1_geo)

PC2_geo <- lm(PC2 ~ elev + lat + lon, data = summary_test)
summary(PC2_geo)

PC3_geo <- lm(PC3 ~ elev + lat + lon, data = summary_test)
summary(PC3_geo)

PC4_geo <- lm(PC4 ~ elev + lat + lon, data = summary_test)
summary(PC4_geo)
```

Takeaway from the correlations is that PC1 and 2 are both correlated with elevation and longitude. Only PC2 is correlated with latitude.

Pc3 and PC4 are not correlated with any geographic variables.

Takeaways from the multiple regressions: 
- PC1:  each of the geographic variables is still a significant predictor of the PC1 values. The R squared is 0.67 which seems pretty good I think. \
- PC2: basically the same as PC1, the R2 is even a little bit higher at 0.71 \
- PC3: no significant predictors, and the R2 value is TINY at 0.04 \
- PC4: basically the same as PC3. \