---
title: "LD_Decay"
author: "Sophie Buysse"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

The goal of this script is to calculate LD decay in my samples. To do so, I need to first calculate LD using plink and then use r to create a graph with the LD output. So, let's start with plink.

This comes up in the GWAS section, so I want to use the same input file that I use for the raw GWAS. the phenotype doesn't really matter and the genotype information for all the GWAS types with the centromere included is all the same.

```{bash, eval = FALSE}
module load PLINK/1.9b_4.1-x86_64

cd $SCRATCH/GWAS

# do for a single chromosome at a time, round one
plink --bfile allSNPs.PlinkFiltering_raw --threads 10 \
--chr 1 --thin 0.5 \
--r2 \
--ld-window 100 \
--ld-window-kb 999999 \
--ld-window-r2 0 \
--allow-no-sex --out LD/allSNPs.PF.chr1

# chr 2
plink --bfile allSNPs.PlinkFiltering_raw --threads 10 \
--chr 2 --thin 0.5 \
--r2 \
--ld-window 100 \
--ld-window-kb 999999 \
--ld-window-r2 0 \
--allow-no-sex --out LD/allSNPs.PF.chr2

# chr 3
plink --bfile allSNPs.PlinkFiltering_raw --threads 10 \
--chr 3 --thin 0.5 \
--r2 \
--ld-window 100 \
--ld-window-kb 999999 \
--ld-window-r2 0 \
--allow-no-sex --out LD/allSNPs.PF.chr3

# chr4
plink --bfile allSNPs.PlinkFiltering_raw --threads 10 \
--chr 4 --thin 0.5 \
--r2 \
--ld-window 100 \
--ld-window-kb 999999 \
--ld-window-r2 0 \
--allow-no-sex --out LD/allSNPs.PF.chr4

# chr5
plink --bfile allSNPs.PlinkFiltering_raw --threads 10 \
--chr 5 --thin 0.5 \
--r2 \
--ld-window 100 \
--ld-window-kb 999999 \
--ld-window-r2 0 \
--allow-no-sex --out LD/allSNPs.PF.chr5

# then get ready to visualize in R
module load GCC/11.2.0  OpenMPI/4.1.1 R/4.2.2
cd LD folder
R

```


Then run this script to plot the result (in R):
```{r}
library(dplyr)
library(ggplot2)

LD_1 <- read.table("allSNPs.PF.chr1.ld", header = T) ##read your .ld data
LD_2 <- read.table("allSNPs.PF.chr2.ld", header = T)
LD_3 <- read.table("allSNPs.PF.chr3.ld", header = T)
LD_4 <- read.table("allSNPs.PF.chr4.ld", header = T)
LD_5 <- read.table("allSNPs.PF.chr5.ld", header = T)
LD <- rbind(LD_1, LD_2, LD_3, LD_4, LD_5)


LD$distancekb <- (LD$BP_B-LD$BP_A)/1000 ## the distance between snp1 and snp2 in kb
max(LD$distance)
#chr 1 this is 317.348
# same value when everything is together

# plot all the points
#plot(x = LD$distancekb, y = LD$R2)

# but we want one with distance bins and an average LD in that bin
# for this function, the second argument is the number of cuts to make, or can give it a vector of cut points
cut_here <- c(seq(from = 0, to = 320, by = 10)) # every 10kb, need to change to part after I know max of everything
LD$grp <- cut(LD$distancekb, cut_here) ## bin 10kb
str(LD)

decay <- LD %>% 
  group_by(grp) %>%
  summarize(r2_mean = mean(R2))

plot(decay$grp, decay$r2_mean)

# group it by chromosome 
decay_chr <- LD %>% 
  group_by(CHR_A) %>%
  group_by(grp, .add = TRUE) %>%
  summarize(r2_mean = mean(R2))

save(decay_chr, file = "/mnt/gs21/scratch/buysseso/GWAS/LD/decay_chr.ROBJ")

```

At this point, moved decay_chr to local machine so I could make higher quality images.

```{r}
library(dplyr)
library(ggplot2)
load("data/decay_chr.ROBJ")

cbPalette <- c("#999999", "#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "#D55E00", "#CC79A7")


ggplot(decay_chr) +
  geom_line(aes(x = grp, y = r2_mean, col = as.factor(CHR_A), group = as.factor(CHR_A)))+
  geom_point(aes(x = grp, y = r2_mean, col = as.factor(CHR_A)))+
  scale_color_manual(name = "Chromosome", values = cbPalette)+
  labs(x = "Distance Between SNPs (kb)", y = "Average R2")+
  theme_classic()+
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5))

ggsave(filename = "LD_decay_chr.png", path = "~/Figures/ManuscriptFigs/supp", height = 7, width = 7, device = "png", dpi = 500)

```