---
title: "SampleSequencingInfo"
author: "Sophia Buysse"
date: "5/20/2022"
output:
  pdf_document: default
  html_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, tidy.opts = list(width.cutoff=75), tidy = TRUE)
```

Goal: identifying sequencing quality by line and by population
Run on the HPCC due to file size, but code is below. Made figures on local machine b/c HPCC down on 8/17

https://gatk.broadinstitute.org/hc/en-us/articles/360035531692-VCF-Variant-Call-Format
.
# Extract Data from vcf File

```{bash, eval = FALSE}
module load GATK
cd /mnt/scratch/buysseso/GVCF

# for centromere included region
gatk VariantsToTable    \
-V allSNPs.noIndelsMafp05MaxMissp75BiallelicQ25Dp5.vcf.gz \
-F CHROM -F POS -F ID -GF GQ -GF AD -GF DP -GF PL -GF GT        \
-O allSNPs.noIndelsMafp05MaxMissp75BiallelicQ25Dp5_SampleInfo

# centromere excluded
gatk VariantsToTable    \
-V NoCent.all.noIndelsMafp05MaxMissp75BiallelicQ25Dp5.recode.vcf \
-F CHROM -F POS -F ID -GF GQ -GF AD -GF DP -GF PL -GF GT        \
-O NoCent.noIndelsMafp05MaxMissp75BiallelicQ25Dp5_SampleInfo
```

ran on terminal, each takes about 3 minutes

# Summarize

```{bash, eval = FALSE}
# load modules
module purge
module load GCC/9.3.0  OpenMPI/4.0.3 R/4.0.3
#module load GDAL/3.0.4-Python-3.8.2
R
```
```{r, eval = FALSE}
library(dplyr)
rm(list = ls())
allSNPs <- read.delim('/mnt/scratch/buysseso/GVCF/allSNPs.noIndelsMafp05MaxMissp75BiallelicQ25Dp5_SampleInfo')
dim(allSNPs)
head(allSNPs[ ,1:6])

NoCent <- read.delim('/mnt/scratch/buysseso/GVCF/NoCent.noIndelsMafp05MaxMissp75BiallelicQ25Dp5_SampleInfo')
dim(NoCent)
head(NoCent[ ,1:6])
```

- ID is a blank column, good to know \ 
- allele depth is the reference allele, alternate allele \
- GT is the gentype \
- don't think PL will be helpful \
- AD is depth for each allele, DP is their sum. use DP b/c most are homozygous if I check random intervals \
\
So all I really want to keep to calculate the means is GQ and DP. Others might be helpful later on, but not right now
```{r, eval = FALSE}
keep.cols.GQ <- seq(from = 4, to = 308, by = 5)
keep.cols.DP <- seq(from = 6, to = 308, by = 5)

# from below to the save line, only keep the one that matches what file was read in
#all.GQ.DP <- allSNPs[,c(1:3, keep.cols.GQ, keep.cols.DP)]
#dim(all.GQ.DP)
#rownames(all.GQ.DP) <- paste0(all.GQ.DP$CHROM, ":", all.GQ.DP$POS)
#head(all.GQ.DP[ ,1:6])
#tmp <- as.data.frame(t(all.GQ.DP))

NC.GQ.DP <- NoCent[,c(1:3, keep.cols.GQ, keep.cols.DP)]
dim(NC.GQ.DP)

rownames(NC.GQ.DP) <- paste0(NC.GQ.DP$CHROM, ":", NC.GQ.DP$POS)
head(NC.GQ.DP[ ,1:6])
tmp <- as.data.frame(t(NC.GQ.DP))
# so now I don't actually need the first three rows b/c the two that matter are incorporated into the column name
tmp <- tmp[4:length(rownames(tmp)), ]
# add population row
tmp$Population <- toupper(substr(rownames(tmp), start = 1, stop = 3))
tmp$Population[tmp$Population == 'ARB'] <- "ARU"
tmp$Population[tmp$Population == 'SPE'] <- "SAL"
ncol(tmp)
# change all to numeric besides last row
factor_numeric <- function(x){
  as.numeric(as.character(x))
}
nrow(tmp)
tmp2 <- apply(tmp[,-nrow(tmp)], 2, factor_numeric)
# add back rownames
rownames(tmp2) <- rownames(tmp)
# using tmp$Population b/c trying to add a tmp2$Population column changes tmp2 to a list instead of a dataframe
Means <- data.frame(ID = rownames(tmp2), Mean = rowMeans(tmp2, na.rm = TRUE), count_na = rowSums(is.na(tmp2)), count_geno = rowSums(!(is.na(tmp2))), Population = tmp$Population)
head(Means)

# save the file that corresponds to what the input file is
#save(Means, file = '/mnt/scratch/buysseso/GVCF/allSNPs_SampleInfo.ROBJ')
save(Means, file = '/mnt/scratch/buysseso/GVCF/NoCent_SampleInfo.ROBJ')
```

Which lines have the lowest quality scores and depth?
```{r}
load ('/mnt/scratch/buysseso/GVCF/NoCent_SampleInfo.ROBJ')
Means[order(Means$Mean), ]
 # lowest depth is ALE16, BAR11, PAL16, RAB4, BIS11: nothing that matches up with the outliers in the PCA

# Lowest GQ is ALE16, PAL16, BAR11, RAB4, BIS11,SPe5, PAN5: matches up with the lowest depth ones but doesn't tell me anything about the PCA, once again.

# most of these also have the fewer total reads according to the fastQC report.
```

okay. so those are line values, but now I want population values, which means I do want dplyr to group by population. I should split up DP and GQ for ease of analysis; try to use the Means dataframe to get the pop values for now.
```{r, eval = FALSE}
rm(list = ls())
load('/mnt/scratch/buysseso/GVCF/allSNPs_SampleInfo.ROBJ')
all <- Means
load('/mnt/scratch/buysseso/GVCF/NoCent_SampleInfo.ROBJ')
NoCent <- Means
rm(Means)

all.GQ <- all[1:61, ]
all_Pop.Sum.GQ <- all.GQ %>% group_by(Population) %>%
  summarize(mean.GQ = mean(Mean), min.GQ = min(Mean), max.GQ = max(Mean), mean.na = mean(count_na), min.na = min(count_na), max.na = max(count_na), mean.geno = mean(count_geno), min.geno = min(count_geno), max.geno = max(count_geno))

nc.GQ <- NoCent[1:61, ]
nc_Pop.Sum.GQ <- nc.GQ %>% group_by(Population) %>%
  summarize(mean.GQ = mean(Mean), min.GQ = min(Mean), max.GQ = max(Mean), mean.na = mean(count_na), min.na = min(count_na), max.na = max(count_na), mean.geno = mean(count_geno), min.geno = min(count_geno), max.geno = max(count_geno))

all.DP <- all[62:122, ]
all_Pop.Sum.DP <- all.DP %>% group_by(Population) %>%
  summarize(mean.DP = mean(Mean), min.DP = min(Mean), max.DP = max(Mean), mean.na = mean(count_na), min.na = min(count_na), max.na = max(count_na), mean.geno = mean(count_geno), min.geno = min(count_geno), max.geno = max(count_geno))

nc.DP <- NoCent[62:122, ]
nc_Pop.Sum.DP <- nc.DP %>% group_by(Population) %>%
  summarize(mean.DP = mean(Mean), min.DP = min(Mean), max.DP = max(Mean), mean.na = mean(count_na), min.na = min(count_na), max.na = max(count_na), mean.geno = mean(count_geno), min.geno = min(count_geno), max.geno = max(count_geno))

```

# Visualize
```{r}
# visualize by line

# all
layout(matrix(c(1,2,3,4,5,6,4,5,6), 3, 3, byrow = TRUE))
hist(all.GQ$Mean)
# quite normal, not super worried about this
hist(all.DP$Mean)
# also quite normal so I am not concerned.
hist(all.GQ$count_na)
#very low value inflated -> good sign!
# I should check which lines have the super high count_na.

plot(Mean ~ count_na, dat = all.GQ, main = "Line.GQ")
plot(Mean ~ count_na, dat = all.DP, main = "Line.DP")
plot(all.GQ$Mean ~ all.DP$Mean)

# these first two plots are almost identical.
# definitely not a one to one line though... but they do make a clear line and are correlated

# No Cent
layout(matrix(c(1,2,3,4,5,6,4,5,6), 3, 3, byrow = TRUE))
hist(nc.GQ$Mean)
# quite normal
hist(nc.DP$Mean)
# super normal
hist(nc.GQ$count_na)
#still very low inflated but a few at higher values, though seemingly lower than the all dataset from the histogram

plot(Mean ~ count_na, dat = nc.GQ, main = "Line.GQ")
plot(Mean ~ count_na, dat = nc.DP, main = "Line.DP")
plot(nc.GQ$Mean ~ nc.DP$Mean)
# first two graphs look less identical than the all dataset. Still super correlated. count_na does have different axis but presumably they are similar?


# visualize by Pop

# all
layout(matrix(c(1,2,3,4,5,6,4,5,6), 3, 3, byrow = TRUE))
hist(all_Pop.Sum.GQ$mean.GQ)
# quite normal yet for the small sample size
hist(all_Pop.Sum.DP$mean.DP)
# hmm. more uniform, not great
hist(all_Pop.Sum.GQ$mean.na)
# hmm. one pop has a higher mean than the rest.
# this pop is BAR, ALE also has second highest mean.na for GQ

# for DP, highest mean.na is for BIS actually. then BAR then PAL, PAN, SAL,ALE, HOR

plot(mean.GQ ~ mean.na, dat = all_Pop.Sum.GQ, main = "Pop.GQ")
plot(mean.DP ~ mean.na, dat = all_Pop.Sum.DP, main = "Pop.DP")
plot(all_Pop.Sum.GQ$mean.GQ ~ all_Pop.Sum.DP$mean.DP)
text(all_Pop.Sum.GQ$mean.GQ ~ all_Pop.Sum.DP$mean.DP, labels = as.character(all_Pop.Sum.GQ$Population))
# still super correlated
# BAR and BIS are super close, 


# NoCent

layout(matrix(c(1,2,3,4,5,6,4,5,6), 3, 3, byrow = TRUE))
hist(nc_Pop.Sum.GQ$mean.GQ)
# pretty uniform
hist(nc_Pop.Sum.DP$mean.DP)
# hmm. more normal but ther eis a gap
hist(nc_Pop.Sum.GQ$mean.na)
# this is actually kinda normal but skewed to lower values just not like zeros
# BAR has the hgihest GQ mean na
# BIS has the highest DP mean na, then BAR

# for DP, highest mean.na is for BIS actually. then BAR then PAL, PAN, SAL,ALE, HOR

plot(mean.GQ ~ mean.na, dat = nc_Pop.Sum.GQ, main = "Pop.GQ")
plot(mean.DP ~ mean.na, dat = nc_Pop.Sum.DP, main = "Pop.DP")
plot(nc_Pop.Sum.GQ$mean.GQ ~ nc_Pop.Sum.DP$mean.DP)
text(nc_Pop.Sum.GQ$mean.GQ ~ nc_Pop.Sum.DP$mean.DP, labels = as.character(nc_Pop.Sum.GQ$Population))
# this looks like identical to the all dataset 
```

So the main takeaway I think is that there aren't huge differences depending on if the centromere is included or not in terms of quality and depth. And there aren't any lines I want to exclude?

These graphs are not saved anywhere. Code is here and files are on the HPCC as noted in this code.